src/cmd/sign.js

```diff
@@ -111,38 +131,64 @@ export default function sign(
         log.warn('No extension ID specified (it will be auto-generated)');
       }
 
-      const signingResult = await signAddon({
+      if (useSubmissionApi && !channel) {
+        throw new UsageError(
+          'channel is a required parameter for the addon submission API'
+        );
+      }
+
+      if (useSubmissionApi && apiProxy) {
+        // https://github.com/mozilla/web-ext/issues/2472
+        throw new UsageError(
+          "apiProxy isn't yet supported for the addon submission API"
+        );
+      }
+
+      if (useSubmissionApi && apiHost.endsWith('/')) {
+        apiHost = apiHost.slice(0, -1);
+      }
+
+      const signSubmitArgs = {
         apiKey,
         apiSecret,
-        apiUrlPrefix,
-        apiProxy,
         timeout,
-        verbose,
         id,
         xpiPath: buildResult.extensionPath,
-        version: manifestData.version,
         downloadDir: artifactsDir,
         channel,
-      });
-
-      if (signingResult.id) {
-        await saveIdToSourceDir(sourceDir, signingResult.id);
-      }
-
-      // All information about the downloaded files would have
-      // already been logged by signAddon().
-      if (signingResult.success) {
-        log.info(`Extension ID: ${signingResult.id}`);
-        log.info('SUCCESS');
-      } else {
+      };
+
+      let result;
+      try {
+        if (useSubmissionApi && channel) {
+          // we verify `channel` is set above, this is just for Flow
+          result = await submitAddon({...signSubmitArgs, apiHost, channel});
+        } else {
+          const { success, ...signAddonResult } = await signAddon({
+            ...signSubmitArgs,
+            apiUrlPrefix,
+            apiProxy,
+            verbose,
+            version: manifestData.version,
+          });
+          if (!success) {
+            throw new Error();
+          }
+          result = signAddonResult;
+        }
+      } catch (clientError) {
         log.info('FAIL');
         throw new WebExtError(
-          'The extension could not be signed');
+          clientError.message || 'The extension could not be signed');
```

Comment:
Is this needed because of the empty error thrown above? If so, maybe that error could use the generic error message.

Reply:
Kind of.  Previously errors that were thrown by `sign-addon` wouldn't have been caught in `sign.js` so the message would propagate; now we're catching all the errors here (as per previous discussion to handle errors in this function, rather than higher up and return `success: false`) so needed to only provide a message for the `success: false` and reuse it otherwise.  I could put that message in the empty error instead, yeah.

Reply:
>  I could put that message in the empty error instead, yeah.

👍